cache/redis_client.go

```diff
@@ -0,0 +1,463 @@
+		c.failures.WithLabelValues(op, reasonServerError)
+		c.failures.WithLabelValues(op, reasonNetworkError)
+		c.failures.WithLabelValues(op, reasonOther)
+	}
+
+	c.skipped = promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+		Name: "redis_operation_skipped_total",
+		Help: "Total number of operations against redis that have been skipped.",
+	}, []string{"operation", "reason"})
+	c.skipped.WithLabelValues(opGetMulti, reasonMaxItemSize)
+	c.skipped.WithLabelValues(opSet, reasonMaxItemSize)
+	c.skipped.WithLabelValues(opSet, reasonAsyncBufferFull)
+
+	c.duration = promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "redis_operation_duration_seconds",
+		Help:    "Duration of operations against redis.",
+		Buckets: []float64{0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.2, 0.5, 1, 3, 6, 10},
+	}, []string{"operation"})
+	c.duration.WithLabelValues(opGetMulti)
+	c.duration.WithLabelValues(opSet)
+	c.duration.WithLabelValues(opDelete)
+
+	c.dataSize = promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+		Name: "redis_operation_data_size_bytes",
+		Help: "Tracks the size of the data stored in and fetched from redis.",
+		Buckets: []float64{
+			32, 256, 512, 1024, 32 * 1024, 256 * 1024, 512 * 1024, 1024 * 1024, 32 * 1024 * 1024, 256 * 1024 * 1024, 512 * 1024 * 1024,
+		},
+	},
+		[]string{"operation"},
+	)
+	c.dataSize.WithLabelValues(opGetMulti)
+	c.dataSize.WithLabelValues(opSet)
+
+	// Start a number of goroutines - processing async operations - equal
+	// to the max concurrency we have.
+	c.workers.Add(c.config.MaxAsyncConcurrency)
+	for i := 0; i < c.config.MaxAsyncConcurrency; i++ {
+		go c.asyncQueueProcessLoop()
+	}
+	return c, nil
+}
+
+// SetAsync implement RemoteCacheClient.
+func (c *redisClient) SetAsync(ctx context.Context, key string, value []byte, ttl time.Duration) error {
+	// Skip hitting redis at all if the item is bigger than the max allowed size.
+	if c.config.MaxItemSize > 0 && uint64(len(value)) > uint64(c.config.MaxItemSize) {
+		c.skipped.WithLabelValues(opSet, reasonMaxItemSize).Inc()
+		return nil
+	}
+
+	err := c.enqueueAsync(func() {
+		start := time.Now()
+
+		_, err := c.Set(ctx, key, value, ttl).Result()
+		if err != nil {
+			level.Debug(c.logger).Log(
+				"msg", "failed to store item to redis",
+				"key", key,
+				"sizeBytes", len(value),
+				"err", err,
+			)
+			c.trackError(opSet, err)
+			return
+		}
+
+		c.dataSize.WithLabelValues(opSet).Observe(float64(len(value)))
+		c.duration.WithLabelValues(opSet).Observe(time.Since(start).Seconds())
+	})
+
+	if errors.Is(err, errMemcachedAsyncBufferFull) {
+		c.skipped.WithLabelValues(opSet, reasonAsyncBufferFull).Inc()
+		level.Debug(c.logger).Log("msg", "failed to store item to redis because the async buffer is full", "err", err, "size", len(c.asyncQueue))
+		return nil
+	}
+	return err
+}
+
+// GetMulti implement RemoteCacheClient.
+func (c *redisClient) GetMulti(ctx context.Context, keys []string) map[string][]byte {
```

Comment:
The same logic in memcached client looks easier to follow. Two questions:
1. Any reason why logic here was written differently?
2. Can we generalize it and reuse the same logic between the two clients?

Reply:
> 1. Any reason why logic here was written differently?

I'd say the main difference between the two implementations is the sharding logic. For memcached this is done on client side whereas for Redis this logic is applied on the cluster instance, when configured in cluster mode. 

> Can we generalize it and reuse the same logic between the two clients?

Perhaps the code contained within the `doWithBatch` function would be susceptible to reuse?

However, it makes use of client-specific interface method for multiple fetch on each implementation (`GetMulti` for memcached and `MGet` for redis), so don't think it's worth it.